std-bits/base/src/main/java/org/enso/base/encoding/ReportingStreamEncoder.java

```diff
@@ -0,0 +1,187 @@
+
+import org.enso.base.Encoding_Utils;
+
+import java.io.BufferedOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.Writer;
+import java.nio.ByteBuffer;
+import java.nio.CharBuffer;
+import java.nio.charset.Charset;
+import java.nio.charset.CharsetEncoder;
+import java.nio.charset.CoderResult;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class ReportingStreamEncoder extends Writer {
+
+  /**
+   * Creates a writer which encodes characters and writes them to the provided output stream.
+   *
+   * <p>The encoder reports any malformed or unmappable characters as problems and replaces them
+   * with the provided replacement sequence.
+   *
+   * <p>The encoder must be closed at the end of the encoding process to indicate that no further
+   * data will be processed so that it can properly handle the finalization of encoding.
+   */
+  public ReportingStreamEncoder(
+      OutputStream outputStream, CharsetEncoder encoder, byte[] replacementSequence) {
+    this.encoder = encoder;
+    bufferedOutputStream = new BufferedOutputStream(outputStream);
+    this.replacementSequence = replacementSequence;
+  }
+
+  private final BufferedOutputStream bufferedOutputStream;
+  private final CharsetEncoder encoder;
+
+  /**
+   * The buffer keeping any input that has already been written but not encoded yet.
+   *
+   * <p>Between the calls to write, it satisfies the invariant that it is in 'reading' mode - to be
+   * able to write to it, it needs to be reallocated, compacted or flipped.
+   */
+  private CharBuffer inputBuffer = CharBuffer.allocate(0);
+
+  private int inputCharactersConsumedBeforeCurrentBuffer = 0;
+
+  private final byte[] replacementSequence;
+
+  private boolean wasClosed = false;
+
+  /**
+   * The buffer re-used for storing encoded output before writing it to the output stream.
+   *
+   * <p>It is cleared after each call to write, so that it can be freshly re-used in the following
+   * call. It is preserved only to avoid re-allocating a big buffer upon each call.
+   */
+  private ByteBuffer outputBuffer = ByteBuffer.allocate(0);
+
+  private void ensureInputBufferHasEnoughFreeSpace(int bytesToAppend) {
+    int freeSpaceInInputBuffer = inputBuffer.capacity() - inputBuffer.remaining();
+
+    // After either compacting the buffer or reallocating it, any remaining input is shifted to
+    // the beginning of the buffer. Thus the bytes that preceded the current position are lost
+    // (because they already have been processed), so we increase the counter to keep the global
+    // position in the input.
+    inputCharactersConsumedBeforeCurrentBuffer += inputBuffer.position();
+
+    if (freeSpaceInInputBuffer < bytesToAppend) {
+      var old = inputBuffer;
+      inputBuffer = CharBuffer.allocate(old.remaining() + bytesToAppend);
+      inputBuffer.put(old);
+    } else {
+      inputBuffer.compact();
```

Comment:
Do you mean the case when `bytesToAppend < inputBuffer.capacity() - inputBuffer.limit() < freeSpaceInInputBuffer`?

We could do that, but that will complicate the logic - because now I need to switch the buffer from read-mode to append-mode. To do so, I'd need to set the new `position` in place of the current `limit` and set the new `limit` for `capacity`. Then we can append to the buffer. And then, after we have appended new input data, we `flip` the buffer back to reading mode. But a `flip` sets the `position` to `0`. But our starting position for reading would need to be the old position from before we prepared the buffer for writing (because at 0 we have stale, already processed data).

We could probably achieve that by setting the buffer's `mark` appropriately and restoring the position after a `flip`. However `compact`ing is a much easier solution - it just moves the pre-existing data to the front, so that after the `flip` we will be able to correctly start reading from `0`.

I'm just not sure if the increased complexity is worth it (as it's always a place for bugs to hide), but if you think it is, I can amend the implementation appropriately.

Reply:
that makes sense.